Add test that `import torch` doesn't modify global logging state

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #87629
* #86213

Fixes https://github.com/pytorch/pytorch/issues/87626

Also adds the same test for `import functorch`. Users have complained at
us when we do modify the global logging state, which has happened in the
past.

Test Plan:
- tested locally; I added `logging.basicConfig` to `torch/__init__.py`
and checked that the test got triggered

zou3519 committed:
Add test that `import torch` doesn't modify global logging state

Fixes https://github.com/pytorch/pytorch/issues/87626

Also adds the same test for `import functorch`. Users have complained at
us when we do modify the global logging state, which has happened in the
past.

Test Plan:
- tested locally; I added `logging.basicConfig` to `torch/__init__.py`
and checked that the test got triggered

[ghstack-poisoned]

albanD commented on test/test_testing.py:
```diff
@@ -1818,6 +1818,29 @@ def test_no_warning_on_import(self) -> None:
             cwd=os.path.dirname(os.path.realpath(__file__)),).decode("utf-8")
         self.assertEquals(out, "")
 
+    @unittest.skipIf(IS_WINDOWS, "importing torch+CUDA on CPU results in warning")
```
No need to skip since you don't check for warnings here?

zou3519 replied:
The warning is going to change the stdout/stderr output, right? Then that affects the string we do a comparison with

albanD commented on test/test_testing.py:
```diff
@@ -1818,6 +1818,29 @@ def test_no_warning_on_import(self) -> None:
+        out = subprocess.check_output(
+            [sys.executable, "-W", "all", "-c", "; ".join(commands)],
+            stderr=subprocess.STDOUT,
+            # On Windows, opening the subprocess with the default CWD makes `import torch`
```
not needed if windows is skipped?

zou3519 replied:
(y)

albanD reviewed (approved):
Sounds good. minor comment about windows

zou3519 committed:
Update on "Add test that `import torch` doesn't modify global logging state"


Fixes https://github.com/pytorch/pytorch/issues/87626

Also adds the same test for `import functorch`. Users have complained at
us when we do modify the global logging state, which has happened in the
past.

Test Plan:
- tested locally; I added `logging.basicConfig` to `torch/__init__.py`
and checked that the test got triggered

[ghstack-poisoned]

zou3519 commented:
Test failure is unrelated (tbh I can't find the failing test in the log, but I verified `test_no_mutate_global_logging_on_import` isn't in the log)

zou3519 commented:
@pytorchbot merge -f "test failures are unrelated"
